File: Serie1PC/TestsSerie1Pc/TestRendezvousChannel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Principal;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Serie1PC;

namespace TestsSerie1Pc
{
    [TestClass]
    public class TestRendezvousChannel
    {

        [TestMethod]
        public void WithOneClientAndOneServerThreadByRunningServerFirst()
        {   
            var synchronizer = new RendezvousChannel<int,int>();
            var waitingToStart = new ManualResetEvent(false);
            var server=new Thread(() =>
            {
                int service;
                waitingToStart.Set();
                var myToken = (RendezvousChannel<int, int>.Token)synchronizer.Accept(Timeout.Infinite, out service);
                synchronizer.Reply(myToken,myToken.service*2);
            });
         
            server.Start();
                int response;
            waitingToStart.WaitOne();
            if (synchronizer.Request(2, Timeout.Infinite, out response))
            {
                Assert.AreEqual(4,response);
            }
        }

        [TestMethod]
        public void WithOneClientAndOneServerThreadByRunningClientFirst()
        {
            var synchronizer = new RendezvousChannel<int, int>();
            var waitingToStart = new ManualResetEvent(false);
            var waitingToEnd=new ManualResetEvent(false);
            var result = 0;
            var client = new Thread(() =>
            {
               
                 int response; 
                waitingToStart.Set();
                synchronizer.Request(2, Timeout.Infinite, out response);
                result = response;
                waitingToEnd.Set();
            });

            client.Start();
            waitingToStart.WaitOne();
                int service;
                var myToken = (RendezvousChannel<int, int>.Token)synchronizer.Accept(Timeout.Infinite, out service);
                synchronizer.Reply(myToken, myToken.service * 2);
            waitingToEnd.WaitOne();
                Assert.AreEqual(4, result);
            
        }
    }
}

File: 1/src/main/java/tests/TestRendezvousChannel.java
package tests;

import static org.junit.Assert.*;

import java.util.concurrent.CountDownLatch;

import org.junit.Assert;
import org.junit.Test;

import concurrency.RendezvousChannel;
import concurrency.RendezvousChannel.Token;


public class TestRendezvousChannel {
	@Test
	public void WithOneClientAndOneServerThreadByRunningServerFirst() throws InterruptedException
	{   
		final RendezvousChannel<Integer, Integer> synchronizer = new RendezvousChannel<Integer, Integer>();
		final CountDownLatch waitingToStart = new CountDownLatch(1);
		Thread server=new Thread(){

			public void run() {
				waitingToStart.countDown();
				Token myToken;
				try {
					myToken = (Token)synchronizer.Accept(Integer.MAX_VALUE);
					synchronizer.Reply(myToken,(Integer)myToken.service*2);
				} catch (InterruptedException e) {/*impossible*/}

			}

		};
		server.start();
		Integer response;
		waitingToStart.await();
		if ((response=(Integer)synchronizer.Request(2, Integer.MAX_VALUE))!=null)
		{

			assertEquals(new Integer(4),response);
		}
	}

	@Test
	public void WithOneClientAndOneServerThreadByRunningClientFirst() throws InterruptedException
	{
		final RendezvousChannel<Integer, Integer> synchronizer = new RendezvousChannel<Integer,Integer>();
		final CountDownLatch waitingToStart = new CountDownLatch(1);
		final CountDownLatch waitingToEnd = new CountDownLatch(1);
		final  Integer result=4;
		Thread client =new Thread(){
			public void run() {
				
				waitingToStart.countDown();
				try {
					Integer response=(Integer)synchronizer.Request(2, Integer.MAX_VALUE);
					assertEquals(result, response);
					waitingToEnd.countDown();
				} catch (InterruptedException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}

			}
		};

		client.start();
		waitingToStart.await();
		int service;
		Token myToken = (Token)synchronizer.Accept(Integer.MAX_VALUE);
		synchronizer.Reply(myToken,(Integer) myToken.service * 2);
		waitingToEnd.await();
		assertEquals(new Integer(4), result);

	}
}



